feat: add COALESCE function

big-andy-coates opened the pull request:
### Description 

The `COALESCE` function takes one or more parameters and returns the first that is non-null.

All parameters must be of the same type.

This function supersedes the existing `IFNULL` function, which only worked for strings and one accepted two parameters.

Note: also added docs for the `ARRAY_LENGH` I added recently, as I seemed to of missed adding docs!

### Testing done 

QTT + unit tests added.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

big-andy-coates commented:
@blueedgenick @rmoff @derekjn @MichaelDrogalis: Just checking from an SQL / Product view that we're cool with this going in?

rmoff commented:
LGTM, thanks Andy!

big-andy-coates commented:
Raised https://github.com/confluentinc/ksql/issues/4831 to track the removal of the deprecated IFNULL.

derekjn commented:
LGTM, solid addition 👍

blueedgenick commented:
Thanks @big-andy-coates , appreciate you soliciting the input. The new COALESCE looks good to me, could benefit from describing in the docs it's behavior with the "non-traditional" i.e. complex datatypes (map, array, struct) - e.g. is a struct containing a null considered to be a null value or not for the purposes of this comparison ? (i'd assume and hope that it's not, but worth calling out and maybe adding tests for). Likewise for an array with a single null entry and so on. These are the situations that may be unfamiliar to a user who is otherwise experienced with using a COALESCE function.

Not keen on removing the IfNull though, i'd much prefer to fix it to handle all types. I know it looks like a subset of COALESCE but it's common in some SQL dialects where COALESCE is not present - it's ok i think to have slightly overlapping functions if they help ease the mental model for users coming from different flavors of RDBMS background. At least MySQL and Vertica, by way of example, have both functions. SQLServer has ISNULL and COALESCE, etc.  

One subtle difference to notice actually is that in many of the common DBMS, COALESCE is actually sugar for a CASE WHEN....CASE WHEN... construct. This is (slightly) significant because it's a "searched case" - meaning that each argument is only evaluated if needed, i.e. if the preceding ones turned out to be null. IFNULL, by contrast, always evaluates both input expressions. Our COALESCE implementation works in a more naive way, having first evaluated all of the expressions, put them in an array, and then passing to the UDF. This is slightly less efficient, especially if the parameters are themselves nested function calls. Not sure if we care about that just yet but it's worth noting for later.

big-andy-coates commented:
@blueedgenick 

I've removed the deprecation, updated the docs and converted #4831 into a task to enhance IFNULL to support more than just STRING args.

derekjn commented:
All good points @blueedgenick, thanks for chiming in 👍

blueedgenick commented:
> @blueedgenick
> 
> I've removed the deprecation, updated the docs and converted #4831 into a task to enhance IFNULL to support more than just STRING args.

Thanks! LGTM

blueedgenick commented on docs-md/developer-guide/ksqldb-reference/scalar-functions.md:
```diff
@@ -186,6 +189,15 @@ The square root of a value.
+Given an array, return the number of elements in the array.
+
+If the supplied parameter is null the method returns 0.
+
```
sorry, i realize i should have seen this `array_length` function going in separately but adding the docs here has called it to my attention :)   - not sure we should return zero for a null input array, that conflicts with the result for an empty array. I would expect that `array_length(null)` returns `NULL`.  From what i can see in a quick google around, that looks to be more common but it's not well documented in many places.

big-andy-coates replied:
>  not sure we should return zero for a null input array, 

I was mulling this over myself.  As you've raised it to... consider it done!

big-andy-coates replied:
Pull the ARRAY_LENGTH stuff out into different PR now that it's growing:

https://github.com/confluentinc/ksql/pull/4848

agavra commented on ksqldb-functional-tests/src/test/resources/query-validation-tests/null.json:
```diff
@@ -76,6 +76,42 @@
+      "name": "coalesce",
+      "statements": [
+        "CREATE STREAM INPUT (COL0 INT KEY, COL1 STRING, COL2 ARRAY<INT>) WITH (kafka_topic='test_topic', value_format='JSON');",
+        "CREATE STREAM OUTPUT AS SELECT COALESCE(COL0, 10, 20) AS A, COALESCE(COL1, 'x') AS B, COALESCE(COL2, ARRAY[10, 20]) AS C FROM INPUT;"
```
can we add a test for `COALESCE(some_column)` (e.g. no constants)

big-andy-coates replied:
done
